src/h3lib/lib/h3Index.c

```diff
@@ -323,75 +323,84 @@ H3Error H3_EXPORT(compactCells)(const H3Index *h3Set, H3Index *compactedSet,
     while (numRemainingHexes) {
         res = H3_GET_RESOLUTION(remainingHexes[0]);
         int parentRes = res - 1;
-        // Put the parents of the hexagons into the temp array
-        // via a hashing mechanism, and use the reserved bits
-        // to track how many times a parent is duplicated
-        for (int i = 0; i < numRemainingHexes; i++) {
-            H3Index currIndex = remainingHexes[i];
-            if (currIndex != 0) {
-                // If the reserved bits were set by the caller, the
-                // algorithm below may encounter undefined behavior
-                // because it expects to have set the reserved bits
-                // itself.
-                if (H3_GET_RESERVED_BITS(currIndex) != 0) {
-                    H3_MEMORY(free)(remainingHexes);
-                    H3_MEMORY(free)(hashSetArray);
-                    return E_CELL_INVALID;
-                }
 
-                H3Index parent;
-                H3Error parentError =
-                    H3_EXPORT(cellToParent)(currIndex, parentRes, &parent);
-                // Should never be reachable as a result of the compact
-                // algorithm. Can happen if cellToParent errors e.g.
-                // because of incompatible resolutions.
-                if (parentError) {
-                    H3_MEMORY(free)(remainingHexes);
-                    H3_MEMORY(free)(hashSetArray);
-                    return parentError;
-                }
-                // Modulus hash the parent into the temp array
-                int loc = (int)(parent % numRemainingHexes);
-                int loopCount = 0;
-                while (hashSetArray[loc] != 0) {
-                    if (loopCount > numRemainingHexes) {  // LCOV_EXCL_BR_LINE
-                        // LCOV_EXCL_START
-                        // This case should not be possible because at most one
-                        // index is placed into hashSetArray per
-                        // numRemainingHexes.
+        // If parentRes is less than zero, we've compacted all the way up to the
+        // base cells. Time to process the remaining cells.
+        if (parentRes >= 0) {
```

Comment:
This seems like a good check to have, although I haven't reviewed this in full because of how complicated this function is. I guess the alternative is breaking at the end of the loop if we get to res 0.

This seems like a good target for reducing how nested the function is, or splitting out into more manageable blocks of code.

Reply:
agree on all points, but I do not volunteer for that refactor <3

Reply:
except breaking at the end of the loop. I'm not sure that's a safe thing to do. Just because this iteration's parentRes is < 0 doesnt mean the next iteration's will be too....I think.  I have no idea, but it seems possible and this way I didn't have to worry about it.

Reply:
Yeah no need to do a full refactor here.